proxy/src/control/cache.rs

```diff
@@ -70,61 +70,82 @@ where
-                    Some(ref old_value) if old_value == &value => {},
-                    // If `insert` returns `Some` with a different value than
-                    // the one we inserted, then we changed the value for that
-                    // key.
-                    Some(_) =>
-                        on_change((key, value), CacheChange::Modification),
+            if inner.get(&key)
+                .map_or(false, |current| current == &new_value) {
+                // The key is already mapped to the inserted value.
+                // No change.
+                return;
+            } else {
+                match inner.insert(key, new_value) {
+                    Some(_) => on_change(CacheChange::Modification {
+                        key,
+                        new_value: inner.get(&key)
+                            .expect("value was just inserted")
+                    }),
                     // If `insert` returns `None`, then there was no old value
                     // previously present. Therefore, we inserted a new value
                     // into the cache.
-                     None => on_change((key, value), CacheChange::Insertion),
+                    None => on_change(CacheChange::Insertion {
+                        key,
+                        value: inner.get(&key)
+                            .expect("value was just inserted")
+                    }),
                 }
             }
         }
 
         if !self.reset_on_next_modification {
-            update_inner(&mut self.inner, iter, on_change);
+            // We don't need to invalidate the cache, so just update
+            // to add the new keys.
+            iter.for_each(|(k, v)| {
+                update_inner(&mut self.inner, k, v, on_change)
+            });
         } else {
-            let to_insert = iter.collect::<IndexMap<K, V>>();
-            update_inner(
-                &mut self.inner,
-                to_insert.iter().map(|(k, v)| (*k, v.clone())),
-                on_change,
-            );
-            self.update_intersection(to_insert, on_change);
+            // The cache was invalidated, so after updating entries present
+            // in `to_insert`, remove any keys not present in `to_insert`.
+            let retained_keys: IndexSet<K> = iter
+                .map(|(k, v)| {
+                    update_inner(&mut self.inner, k, v, on_change);
+                    k
+                })
+                .collect();
+            // This retain is similar to `update_intersection`'s removal of
+            // non-present keys but without updating values, since we've
+            // already updated any values present in `to_insert`. We do this
+            // instead of just calling `update_intersection` so we don't have
+            // to clone the values we already know were inserted here.
+            self.inner.retain(|key, _| if retained_keys.contains(key) {
```

Comment:
There was previously a call to `update_intersection` here. I've replaced it with this code, because `update_intersection` takes an `IndexMap<K, V>`, which would require us to clone the inserted values. However, all we care about here is removing any stale keys, regardless of what their values are, since any changed values were already updated by `update_inner`, so we only need to filter out keys that weren't in the inserted key-value pairss.